core-relations/src/free_join/mod.rs

```diff
@@ -529,21 +544,39 @@ impl Database {
         ever_changed
     }
 
+    /// A "fast path" merge method that is not optimized for parallelism and does not respect read
+    /// and write dependencies. This ends up being faster than the full "strata-aware" option in
+    /// the body of `merge_all`.
+    fn merge_simple(&mut self, mut to_merge: SmallVec<[TableId; 4]>) -> bool {
+        let mut changed = false;
+        while !to_merge.is_empty() {
+            for table_id in to_merge.iter().copied() {
+                let mut info = self.tables.unwrap_val(table_id);
+                let mut es = ExecutionState::new(self.read_only_view(), Default::default());
+                changed |= info.table.merge(&mut es).added || es.changed;
+                self.tables.insert(table_id, info);
+            }
+            to_merge = self.notification_list.reset();
+        }
+        changed
+    }
+
     /// A low-level helper for merging pending updates to a particular function.
     ///
     /// Callers should prefer `merge_all`, as the process of merging the data
     /// for a particular table may cause other updates to be buffered
     /// elesewhere. The `merge_all` method runs merges to a fixed point to avoid
     /// surprises here.
-    pub fn merge_table(&mut self, table: TableId) {
+    pub fn merge_table(&mut self, table: TableId) -> bool {
         let mut info = self.tables.unwrap_val(table);
         self.total_size_estimate = self.total_size_estimate.wrapping_sub(info.table.len());
-        let _table_changed = info.table.merge(&mut ExecutionState::new(
+        let table_changed = info.table.merge(&mut ExecutionState::new(
             self.read_only_view(),
             Default::default(),
         ));
         self.total_size_estimate = self.total_size_estimate.wrapping_add(info.table.len());
         self.tables.insert(table, info);
+        table_changed.added || table_changed.removed
```

Comment:
One thing I'm curious about is that merge_all and merge_simple only count `added` as an update, while here we count both `added` and `removed`. Also, is the returned flag ever used currently?

Reply:
Yeah... I think just `added` is needed? I think this is a subtle question. You can definitely write rules that reach a different fixed point based on `removed` by writing an `ExternalFunction` that does something based on a row not being present in a table. But such a rule would not be legal egglog as it would allow for the encoding of non-stratified negation / break unique fixed points. I'll remove the `removed` check for now